### PR TITLE
Fix WaitEventSetWait timeout in unified executor

### PR DESCRIPTION
**Pull request by marcocitus**
We currently determine the `WaitEventSetWait` timeout using `DeadlineTimestampTzToTimeout` where the deadline is the start time of `RunDistributedExecution` plus the node connection timeout. However, this has several issues a) the `RunDistributedExecution` timeout may not be aligned with the actual worker pool timeout, b) the timeout value converges to <= 0 which effectively disable the timeout (i.e. we might not wake up when the pool actually times out) and c) it ignores node connection timeouts that start later (after a failover).

This PR fixes the issue by going through all worker pools and determining the number of milliseconds until the connection timeout. There could be a slightly simpler solution, but this one can easily be extended to consider other time-based events such as slow start.

### CODE REVIEW

File: src/backend/distributed/executor/executor.c [hunk 1]
```diff
@@ -1658,36 +1665,110 @@ CheckConnectionTimeout(WorkerPool *workerPool)
...
+			ereport(logLevel, (errcode(ERRCODE_CONNECTION_FAILURE),
+							   errmsg("could not establish any connections to the node "
+									  "%s:%d after %u ms", workerPool->node->workerName,
+									  workerPool->node->workerPort,
+									  NodeConnectionTimeout)));
+		}
+		else
+		{
+			/* stop interrupting WaitEventSetWait for timeouts */
+			workerPool->survivedTimeout = true;
```
Comment (onderkalaci) [inline src/backend/distributed/executor/executor.c]: Totally optional: I find slightly hard to follow the calculation and the name of `survivedTimeout `. It might be easier to read if we rename the flag to `checkForPoolTimeout` (or such) and revert the value of the flag.

In that case, we could even use that flag to early return in this function. Because with this change we're actually executing some of the checks of the function more often.

File: src/backend/distributed/executor/executor.c [hunk 2]
```diff
@@ -1524,12 +1525,18 @@ ManageWorkerPool(WorkerPool *workerPool)
...
+	if (workerPool->failed)
+	{
+		/* connection pool failed */
+		return;
+	}
+
 	/* we might fail the execution or warn the user about connection timeouts */
 	CheckConnectionTimeout(workerPool);
 
 	if (failedConnectionCount >= 1)
```
Comment (onderkalaci) [inline src/backend/distributed/executor/executor.c]: shall we move this check above
```C

if (workerPool->failed || failedConnectionCount >= 1)
{
    return
}
```

Comment (marcocitus) [reply src/backend/distributed/executor/executor.c]: I figured there might be situations in which we want to check for timeout after one connection has failed, e.g. we open 4 connections and 1 fails and the second never responds.
